README.md
# diqwest

This crate extends `reqwest` to be able to send requests with digest auth flow.

When you send a request with digest auth flow this first request will be executed. In case the response is a `401` the `www-authenticate` header is parsed and the answer is calculated. The initial request is executed again with additional `Authorization` header. The response will be returned from `send_with_digest_auth()`.

In case the first response is not a `401` this first response is returned from `send_with_digest_auth()` without any manipulation. In case the first response is a `401` but the `www-authenticate` header is missing the first reponse is returned as well.

## Examples

### Async (default)
```rust
use diqwest::WithDigestAuth;
use reqwest::{Client, Response};

// Call `.send_with_digest_auth()` on `RequestBuilder` like `send()`
let response: Response = Client::new()
  .get("url")
  .send_with_digest_auth("username", "password")
  .await?;
```

### Blocking (feature flag `blocking` has to be enabled in `Cargo.toml`)

```rust
use diqwest::blocking::WithDigestAuth;
use reqwest::blocking::{Client, Response};

// Call `.send_with_digest_auth()` on `RequestBuilder` like `send()`
let response: Response = Client::new()
  .get("url")
  .send_with_digest_auth("username", "password")?;
```


CHANGELOG.md
# 1.0.1

- bugfix: the passed uri to the www-athenticate header must only consist of the path of the uri

# 1.0.0

- establish usage with `blocking` behavior as a non default feature
- the package structure changed in a breaking way, sematic version for major +1

# 0.1.2

- get rid of explicit `http 0.2` dependency and use it transitive from `reqwest` 

# 0.1.1

- make the dependencies less restrictive

# 0.1.0

- initial version
- add async functionality to extend `reqwest` to send requests with digest auth flow 


src/core.rs
use async_trait::async_trait;
use digest_auth::{AuthContext, HttpMethod};
use reqwest::{RequestBuilder, Response, StatusCode};

use Error::RequestBuilderNotCloneableError;

use crate::error::{Error, Result};

/// A trait to extend the functionality of a `RequestBuilder` to send a request with digest auth flow.
///
/// Call it at the end of your `RequestBuilder` chain like you would use `send()`.
#[async_trait]
pub trait WithDigestAuth {
  async fn send_with_digest_auth(&self, username: &str, password: &str) -> Result<Response>;
}

#[async_trait]
impl WithDigestAuth for RequestBuilder {
  async fn send_with_digest_auth(&self, username: &str, password: &str) -> Result<Response> {
    fn clone_request_builder(request_builder: &RequestBuilder) -> std::result::Result<RequestBuilder, Error> {
      request_builder.try_clone().ok_or(RequestBuilderNotCloneableError)
    }

    let first_response = clone_request_builder(self)?.send().await?;
    match first_response.status() {
      StatusCode::UNAUTHORIZED => {
        let request = clone_request_builder(self)?.build()?;
        let url = request.url();
        let method = HttpMethod::from(request.method().as_str());
        let body = request.body().and_then(|b| b.as_bytes());
        let answer =
          DigestAuthHelper::parse_digest_auth_header(first_response, url.as_str(), method, body, username, password)?;

        Ok(clone_request_builder(self)?.header("Authorization", answer).send().await?)
      }
      _ => Ok(first_response),
    }
  }
}

struct DigestAuthHelper;

impl DigestAuthHelper {
  fn parse_digest_auth_header(
    response: Response,
    uri: &str,
    method: HttpMethod,
    body: Option<&[u8]>,
    public: &str,
    private: &str,
  ) -> Result<String> {
    let www_auth = response.headers()["www-authenticate"].to_str()?;
    let context = AuthContext::new_with_method(public, private, uri, body, method);
    let mut prompt = digest_auth::parse(www_auth)?;

    Ok(prompt.respond(&context)?.to_header_string())
  }
}


src/error.rs
use std::fmt::{Debug, Display, Formatter};
use std::result;

use reqwest::header::ToStrError;

#[derive(Debug)]
pub enum Error {
  Reqwest(reqwest::Error),
  DigestAuth(digest_auth::Error),
  ToStr(reqwest::header::ToStrError),
  AuthHeaderMissing,
  RequestBuilderNotCloneable,
}

pub type Result<T> = result::Result<T, Error>;

impl Display for Error {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    match self {
      Error::Reqwest(e) => std::fmt::Display::fmt(e, f),
      Error::DigestAuth(e) => std::fmt::Display::fmt(e, f),
      Error::ToStr(e) => std::fmt::Display::fmt(e, f),
      Error::RequestBuilderNotCloneable => write!(f, "Request body must not be a stream."),
      Error::AuthHeaderMissing => write!(f, "The header 'www-authenticate' is missing."),
    }
  }
}

impl std::error::Error for Error {}

impl From<reqwest::Error> for Error {
  fn from(e: reqwest::Error) -> Self {
    Error::Reqwest(e)
  }
}

impl From<digest_auth::Error> for Error {
  fn from(e: digest_auth::Error) -> Self {
    Error::DigestAuth(e)
  }
}

impl From<reqwest::header::ToStrError> for Error {
  fn from(e: ToStrError) -> Self {
    Error::ToStr(e)
  }
}


src/blocking.rs
use digest_auth::HttpMethod;
use reqwest::blocking::{RequestBuilder, Response};
use reqwest::StatusCode;

use crate::error::Error::{AuthHeaderMissing, RequestBuilderNotCloneable};
use crate::error::Result;
use crate::parse_digest_auth_header;

/// A trait to extend the functionality of a blocking `RequestBuilder` to send a request with digest auth flow.
///
/// Call it at the end of your `RequestBuilder` chain like you would use `send()`.
pub trait WithDigestAuth {
  fn send_with_digest_auth(&self, username: &str, password: &str) -> Result<Response>;
}

impl WithDigestAuth for RequestBuilder {
  fn send_with_digest_auth(&self, username: &str, password: &str) -> Result<Response> {
    fn clone_request_builder(request_builder: &RequestBuilder) -> Result<RequestBuilder> {
      request_builder.try_clone().ok_or(RequestBuilderNotCloneable)
    }

    let first_response = clone_request_builder(self)?.send()?;
    match first_response.status() {
      StatusCode::UNAUTHORIZED => {
        let request = clone_request_builder(self)?.build()?;
        let path = request.url().path();
        let method = HttpMethod::from(request.method().as_str());
        let body = request.body().and_then(|b| b.as_bytes());
        let answer = parse_digest_auth_header(first_response.headers(), path, method, body, username, password);

        match answer {
          Ok(answer) => Ok(
            clone_request_builder(self)?
              .header("Authorization", answer.to_header_string())
              .send()?,
          ),
          Err(AuthHeaderMissing) => Ok(first_response),
          Err(error) => Err(error),
        }
      }
      _ => Ok(first_response),
    }
  }
}

#[cfg(test)]
mod tests {
  use crate::blocking::WithDigestAuth;
  use crate::parse_digest_auth_header;

  use digest_auth::HttpMethod;
  use mockito::{mock, Mock};
  use reqwest::blocking::Client;
  use reqwest::{
    header::{HeaderMap, HeaderValue},
    StatusCode,
  };

  #[test]
  fn given_non_digest_auth_endpoint_when_send_with_da_then_request_executed_normally() {
    // Given I have a GET request against a non digest auth endpoint
    let path = "/test";
    let mock = mock("GET", path).with_status(200).create();
    let request = Client::new().get(format!("{domain}{path}", domain = mockito::server_url()));

    // When I send with digest auth
    let response = request.send_with_digest_auth("username", "password").unwrap();

    // Then the response's status is OK
    Mock::assert(&mock);
    assert_eq!(&response.status(), &StatusCode::OK);
  }

  #[test]
  fn given_non_digest_auth_endpoint_unauthorized_when_send_with_da_then_request_fails_with_401() {
    // Given I have a GET request against a non digest auth  but authorized endpoint
    let path = "/test";
    let mock = mock("GET", path).with_status(401).create();
    let request = Client::new().get(format!("{domain}{path}", domain = mockito::server_url()));

    // When I send with digest auth
    let response = request.send_with_digest_auth("username", "password").unwrap();

    // Then the response's final status is UNAUTHORIZED
    Mock::assert(&mock);
    assert_eq!(&response.status(), &StatusCode::UNAUTHORIZED);
  }

  #[test]
  fn given_digest_auth_endpoint_authorized_when_send_with_da_then_request_succeeds() {
    // Given I have a GET request against a digest auth endpoint with valid 'www-authenticate' header
    let path = "/test";
    let url = format!("{domain}{path}", domain = mockito::server_url());
    let www_authenticate = "Digest realm=\"testrealm@host.com\",qop=\"auth,auth-int\",nonce=\"dcd98b7102dd2f0e8b11d0f600bfb0c093\",opaque=\"5ccc069c403ebaf9f0171e9517f40e41\"";
    let mut header = HeaderMap::new();
    header.insert("www-authenticate", HeaderValue::from_static(www_authenticate));
    let auth_header = parse_digest_auth_header(&header, path, HttpMethod::GET, None, "username", "password").unwrap();

    let first_request = mock("GET", path)
      .with_status(401)
      .with_header("www-authenticate", www_authenticate)
      .create();
    let second_request = mock("GET", path)
      .with_header("Authorization", &auth_header.to_header_string())
      .with_status(200)
      .create();

    let request = Client::new().get(&url);

    // When I send with digest auth
    let response = request.send_with_digest_auth("username", "password").unwrap();

    // Then the response's final status is OK
    Mock::assert(&first_request);
    Mock::assert(&second_request);
    assert_eq!(&response.status(), &StatusCode::OK);
  }
}


src/lib.rs
//! This crate extends `reqwest` to be able to send requests with digest auth flow.
//!
//! When you send a request with digest auth flow this first request will be executed. In case
//! the response is a `401` the `www-authenticate` header is parsed and the answer is calculated.
//! The initial request is executed again with additional `Authorization` header. The response
//! will be returned from `send_with_digest_auth()`.
//!
//! In case the first response is not a `401` this first response is returned from
//! `send_with_digest_auth()` without any manipulation. In case the first response is a `401`
//! but the `www-authenticate` header is missing the first reponse is returned as well.
//!
//! # Examples
//!
//! By default this crate works async:
//!
//! ```compile_fail
//! use diqwest::WithDigestAuth;
//! use reqwest::{Client, Response};
//!
//! // Call `.send_with_digest_auth()` on `RequestBuilder` like `send()`
//! let response: Response = Client::new()
//!   .get("url")
//!   .send_with_digest_auth("username", "password")
//!   .await?;
//!
//! ```
//!
//! In case you need blocking behavior enable the `blocking` feature in your `Cargo.toml`.
//!
//! ```compile_fail
//! use diqwest::blocking::WithDigestAuth;
//! use reqwest::blocking::{Client, Response};
//!
//! // Call `.send_with_digest_auth()` on `RequestBuilder` like `send()`
//! let response: Response = Client::new()
//!   .get("url")
//!   .send_with_digest_auth("username", "password")?;
//!
//! ```
//!

#[cfg(feature = "blocking")]
pub mod blocking;
pub mod error;

use async_trait::async_trait;
use digest_auth::{AuthContext, AuthorizationHeader, HttpMethod};
use error::Error;
use reqwest::header::HeaderMap;
use reqwest::{RequestBuilder, Response, StatusCode};

use crate::error::Error::RequestBuilderNotCloneable;
use crate::error::Result;

/// A trait to extend the functionality of an async `RequestBuilder` to send a request with digest auth flow.
///
/// Call it at the end of your `RequestBuilder` chain like you would use `send()`.
#[async_trait]
pub trait WithDigestAuth {
  async fn send_with_digest_auth(&self, username: &str, password: &str) -> Result<Response>;
}

#[async_trait]
impl WithDigestAuth for RequestBuilder {
  async fn send_with_digest_auth(&self, username: &str, password: &str) -> Result<Response> {
    fn clone_request_builder(request_builder: &RequestBuilder) -> Result<RequestBuilder> {
      request_builder.try_clone().ok_or(RequestBuilderNotCloneable)
    }

    let first_response = clone_request_builder(self)?.send().await?;
    match first_response.status() {
      StatusCode::UNAUTHORIZED => {
        let request = clone_request_builder(self)?.build()?;
        let path = request.url().path();
        let method = HttpMethod::from(request.method().as_str());
        let body = request.body().and_then(|b| b.as_bytes());
        let answer = parse_digest_auth_header(first_response.headers(), path, method, body, username, password);

        match answer {
          Ok(answer) => Ok(
            clone_request_builder(self)?
              .header("Authorization", answer.to_header_string())
              .send()
              .await?,
          ),
          Err(error::Error::AuthHeaderMissing) => Ok(first_response),
          Err(error) => Err(error),
        }
      }
      _ => Ok(first_response),
    }
  }
}

fn parse_digest_auth_header(
  header: &HeaderMap,
  path: &str,
  method: HttpMethod,
  body: Option<&[u8]>,
  username: &str,
  password: &str,
) -> Result<AuthorizationHeader> {
  let www_auth = header.get("www-authenticate").ok_or(Error::AuthHeaderMissing)?.to_str()?;
  let context = AuthContext::new_with_method(username, password, path, body, method);
  let mut prompt = digest_auth::parse(www_auth)?;

  Ok(prompt.respond(&context)?)
}

#[cfg(test)]
mod tests {
  use crate::{parse_digest_auth_header, WithDigestAuth};

  use digest_auth::HttpMethod;
  use mockito::{mock, Mock};
  use reqwest::{
    header::{HeaderMap, HeaderValue},
    Client, StatusCode,
  };

  #[tokio::test]
  async fn given_non_digest_auth_endpoint_when_send_with_da_then_request_executed_normally() {
    // Given I have a GET request against a non digest auth endpoint
    let path = "/test";
    let mock = mock("GET", path).with_status(200).create();
    let request = Client::new().get(format!("{domain}{path}", domain = mockito::server_url()));

    // When I send with digest auth
    let response = request.send_with_digest_auth("username", "password").await.unwrap();

    // Then the response's status is OK
    Mock::assert(&mock);
    assert_eq!(&response.status(), &StatusCode::OK);
  }

  #[tokio::test]
  async fn given_non_digest_auth_endpoint_unauthorized_when_send_with_da_then_request_fails_with_401() {
    // Given I have a GET request against a non digest auth  but authorized endpoint
    let path = "/test";
    let mock = mock("GET", path).with_status(401).create();
    let request = Client::new().get(format!("{domain}{path}", domain = mockito::server_url()));

    // When I send with digest auth
    let response = request.send_with_digest_auth("username", "password").await.unwrap();

    // Then the response's final status is UNAUTHORIZED
    Mock::assert(&mock);
    assert_eq!(&response.status(), &StatusCode::UNAUTHORIZED);
  }

  #[tokio::test]
  async fn given_digest_auth_endpoint_authorized_when_send_with_da_then_request_succeeds() {
    // Given I have a GET request against a digest auth endpoint with valid 'www-authenticate' header
    let path = "/test";
    let url = format!("{domain}{path}", domain = mockito::server_url());
    let www_authenticate = "Digest realm=\"testrealm@host.com\",qop=\"auth,auth-int\",nonce=\"dcd98b7102dd2f0e8b11d0f600bfb0c093\",opaque=\"5ccc069c403ebaf9f0171e9517f40e41\"";
    let mut header = HeaderMap::new();
    header.insert("www-authenticate", HeaderValue::from_static(www_authenticate));
    let auth_header = parse_digest_auth_header(&header, path, HttpMethod::GET, None, "username", "password").unwrap();

    let first_request = mock("GET", path)
      .with_status(401)
      .with_header("www-authenticate", www_authenticate)
      .create();
    let second_request = mock("GET", path)
      .with_header("Authorization", &auth_header.to_header_string())
      .with_status(200)
      .create();

    let request = Client::new().get(url);

    // When I send with digest auth
    let response = request.send_with_digest_auth("username", "password").await.unwrap();

    // Then the response's final status is OK
    Mock::assert(&first_request);
    Mock::assert(&second_request);
    assert_eq!(&response.status(), &StatusCode::OK);
  }
}
